File: src/main/java/com/trainingplatform/application/services/AuthService.java
package com.trainingplatform.application.services;

import com.trainingplatform.core.entities.User;
import com.trainingplatform.core.entities.Rep;
import com.trainingplatform.infrastructure.repositories.UserRepository;
import com.trainingplatform.infrastructure.repositories.RepRepository;
import com.trainingplatform.application.dto.LoginRequest;
import com.trainingplatform.application.dto.RegisterRequest;
import com.trainingplatform.application.dto.AuthResponse;
import com.trainingplatform.infrastructure.security.JwtTokenProvider;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
public class AuthService {
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private RepRepository repRepository;
    
    @Autowired
    private PasswordEncoder passwordEncoder;
    
    @Autowired
    private JwtTokenProvider jwtTokenProvider;
    
    @Autowired
    private AuthenticationManager authenticationManager;

    public AuthResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
            new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        User user = userRepository.findByEmail(request.getEmail())
            .orElseThrow(() -> new RuntimeException("User not found"));

        String token = jwtTokenProvider.generateToken(authentication);
        
        Rep repProfile = repRepository.findByUserId(user.getId()).orElse(null);

        return new AuthResponse(token, user, repProfile);
    }

    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already exists");
        }

        User user = new User();
        user.setName(request.getName());
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        user.setRole(request.getRole());
        user.setDepartment(request.getDepartment());
        
        Map<String, Object> aiProfile = new HashMap<>();
        aiProfile.put("strengths", new String[]{});
        aiProfile.put("improvementAreas", new String[]{});
        aiProfile.put("preferredLearningPace", "medium");
        aiProfile.put("motivationFactors", new String[]{});
        user.setAiPersonalityProfile(aiProfile);

        user = userRepository.save(user);

        // Create rep profile
        Rep rep = new Rep();
        rep.setUserId(user.getId());
        rep.setName(user.getName());
        rep.setEmail(user.getEmail());
        rep.setRole(user.getRole());
        rep.setDepartment(user.getDepartment());
        rep.setAiPersonalityProfile(aiProfile);
        
        rep = repRepository.save(rep);

        Authentication authentication = authenticationManager.authenticate(
            new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        String token = jwtTokenProvider.generateToken(authentication);

        return new AuthResponse(token, user, rep);
    }

    public User getCurrentUser(String email) {
        return userRepository.findByEmail(email)
            .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public Rep getRepProfile(String userId) {
        return repRepository.findByUserId(userId)
            .orElseThrow(() -> new RuntimeException("Rep profile not found"));
    }
}

File: src/services/authService.ts
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import User from '../models/User';
import Rep from '../models/Rep';
import { AppError } from '../middleware/errorHandler';

export interface LoginRequest {
  email: string;
  password: string;
}

export interface RegisterRequest {
  name: string;
  email: string;
  password: string;
  role?: string;
  department?: string;
}

export interface AuthResponse {
  token: string;
  user: {
    id: string;
    name: string;
    email: string;
    role: string;
    department?: string;
  };
  repProfile?: any;
}

class AuthService {
  private generateToken(userId: string, email: string, role: string): string {
    const jwtSecret = process.env.JWT_SECRET || 'mySecretKey';
    const jwtExpiration = process.env.JWT_EXPIRATION || '86400000';

    return jwt.sign(
      { id: userId, email, role },
      jwtSecret,
      { expiresIn: parseInt(jwtExpiration) / 1000 }
    );
  }

  async login(loginData: LoginRequest): Promise<AuthResponse> {
    const { email, password } = loginData;

    const user = await User.findOne({ email }).select('+password');
    if (!user) {
      throw new AppError('Invalid email or password', 401);
    }

    const isPasswordValid = await bcrypt.compare(password, user.password);
    if (!isPasswordValid) {
      throw new AppError('Invalid email or password', 401);
    }

    const token = this.generateToken(user._id.toString(), user.email, user.role);

    const repProfile = await Rep.findOne({ userId: user._id.toString() });

    return {
      token,
      user: {
        id: user._id.toString(),
        name: user.name,
        email: user.email,
        role: user.role,
        department: user.department
      },
      repProfile: repProfile || undefined
    };
  }

  async register(registerData: RegisterRequest): Promise<AuthResponse> {
    const { name, email, password, role, department } = registerData;

    const existingUser = await User.findOne({ email });
    if (existingUser) {
      throw new AppError('Email already exists', 400);
    }

    const hashedPassword = await bcrypt.hash(password, 10);

    const aiProfile = {
      strengths: [],
      improvementAreas: [],
      preferredLearningPace: 'medium',
      motivationFactors: []
    };

    const user = await User.create({
      name,
      email,
      password: hashedPassword,
      role: role || 'trainee',
      department,
      aiPersonalityProfile: aiProfile
    });

    const rep = await Rep.create({
      userId: user._id.toString(),
      name: user.name,
      email: user.email,
      role: user.role,
      department: user.department,
      aiPersonalityProfile: aiProfile
    });

    const token = this.generateToken(user._id.toString(), user.email, user.role);

    return {
      token,
      user: {
        id: user._id.toString(),
        name: user.name,
        email: user.email,
        role: user.role,
        department: user.department
      },
      repProfile: rep
    };
  }

  async getCurrentUser(email: string) {
    const user = await User.findOne({ email });
    if (!user) {
      throw new AppError('User not found', 404);
    }
    return user;
  }

  async getRepProfile(userId: string) {
    const rep = await Rep.findOne({ userId });
    if (!rep) {
      throw new AppError('Rep profile not found', 404);
    }
    return rep;
  }
}

export default new AuthService();
